src/check.rs

```diff
@@ -1,37 +1,66 @@
 //! Provides functionality for checking the availablility of URLs.
-use std::{fmt, path::PathBuf};
+use std::collections::HashSet;
+use std::fmt;
+use std::fs::read_to_string;
+use std::path::{Path, PathBuf};
 
 use log::debug;
+use once_cell::sync::Lazy;
+use regex::Regex;
 use url::Url;
 
+use cached::cached_key_result;
+use cached::SizedCache;
+
 use super::CheckContext;
 
+use crate::parse::parse_fragments;
+
 const PREFIX_BLACKLIST: [&str; 1] = ["https://doc.rust-lang.org"];
 
 #[derive(Debug)]
-pub enum HttpError {
-    UnexpectedStatus(Url, ureq::Response),
-    Fetch(Url, ureq::Error),
+pub enum IoError {
+    HttpUnexpectedStatus(Url, ureq::Response),
+    HttpFetch(Url, ureq::Error),
+    FileIo(String, std::io::Error),
 }
 
-impl fmt::Display for HttpError {
+impl fmt::Display for IoError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
-            HttpError::UnexpectedStatus(url, resp) => write!(
+            IoError::HttpUnexpectedStatus(url, resp) => write!(
                 f,
                 "Unexpected HTTP status fetching {}: {}",
                 url,
                 resp.status_text()
             ),
-            HttpError::Fetch(url, e) => write!(f, "Error fetching {}: {}", url, e),
+            IoError::HttpFetch(url, e) => write!(f, "Error fetching {}: {}", url, e),
+            IoError::FileIo(url, e) => write!(f, "Error fetching {}: {}", url, e),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Link {
+    File(String),
+    Http(Url),
+}
+
+impl fmt::Display for Link {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Link::File(path) => f.write_str(path),
+            Link::Http(url) => f.write_str(url.as_str()),
         }
     }
 }
 
 #[derive(Debug)]
 pub enum CheckError {
     File(PathBuf),
-    Http(Box<HttpError>),
+    Http(Url),
+    Fragment(Link, String, Option<Vec<String>>),
+    Io(Box<IoError>),
```

Comment:
```suggestion
    /// A linked HTTP URL did not exist
    Http(Url),
    /// The linked file existed, but was missing the linked HTML anchor
    Fragment(Link, String, Option<Vec<String>>),
    /// An error occured while trying to find whether the file or URL existed
    Io(Box<IoError>),
```
and please add one for `File` too, github's `suggestion` button doesn't like deleted lines.